renderdoc/driver/metal/metal_device.cpp

```diff
@@ -38,3 +51,121 @@ MTL::Device *WrappedMTLDevice::MTLCreateSystemDefaultDevice(MTL::Device *realMTL
 
   return UnwrapObjC<MTL::Device *>(wrappedMTLDevice);
 }
+
+MTL::Library *WrappedMTLDevice::newDefaultLibrary()
+{
+  MTL::Library *realMTLLibrary;
+
+  SERIALISE_TIME_CALL(realMTLLibrary = Unwrap<MTL::Device *>(this)->newDefaultLibrary());
+  MetalResourceManager::UnwrapHelper<MTL::Library *>::Outer *wrappedMTLLibrary;
+  ResourceId id = GetResourceManager()->WrapResource(realMTLLibrary, wrappedMTLLibrary);
+  if(IsCaptureMode(m_State))
+  {
+    void *pData;
+    uint32_t bytesCount;
+    ObjC::Get_defaultLibraryData(pData, bytesCount);
+    Chunk *chunk = NULL;
+    {
+      CACHE_THREAD_SERIALISER();
+      SCOPED_SERIALISE_CHUNK(MetalChunk::MTLDevice_newDefaultLibrary);
+      Serialise_mtlDevice_newDefaultLibrary(ser, this, wrappedMTLLibrary, pData, bytesCount);
+      chunk = scope.Get();
+    }
+    free(pData);
+    MetalResourceRecord *record = GetResourceManager()->AddResourceRecord(wrappedMTLLibrary);
+    record->AddChunk(chunk);
+    GetResourceManager()->MarkResourceFrameReferenced(id, eFrameRef_Read);
+  }
+  else
+  {
+    // TODO: implement RD MTL replay
+    //     GetResourceManager()->AddLiveResource(id, wrappedMTLLibrary);
+  }
+  return UnwrapObjC<MTL::Library *>(wrappedMTLLibrary);
+}
+
+MTL::Library *WrappedMTLDevice::newLibraryWithSource(NS::String *source,
+                                                     MTL::CompileOptions *options, NS::Error **error)
+{
+  MTL::Library *realMTLLibrary;
+  SERIALISE_TIME_CALL(realMTLLibrary =
+                          Unwrap<MTL::Device *>(this)->newLibrary(source, options, error));
+  MetalResourceManager::UnwrapHelper<MTL::Library *>::Outer *wrappedMTLLibrary;
+  ResourceId id = GetResourceManager()->WrapResource(realMTLLibrary, wrappedMTLLibrary);
+  if(IsCaptureMode(m_State))
+  {
+    Chunk *chunk = NULL;
+    {
+      CACHE_THREAD_SERIALISER();
+      SCOPED_SERIALISE_CHUNK(MetalChunk::MTLDevice_newLibraryWithSource);
+      Serialise_mtlDevice_newLibraryWithSource(ser, this, wrappedMTLLibrary, source, options);
+      chunk = scope.Get();
+    }
+    MetalResourceRecord *record = GetResourceManager()->AddResourceRecord(wrappedMTLLibrary);
+    record->AddChunk(chunk);
+    GetResourceManager()->MarkResourceFrameReferenced(id, eFrameRef_Read);
+  }
+  else
+  {
+    // TODO: implement RD MTL replay
+    //     GetResourceManager()->AddLiveResource(id, wrappedMTLLibrary);
+  }
+  return UnwrapObjC<MTL::Library *>(wrappedMTLLibrary);
+}
+
+template <typename SerialiserType>
+bool WrappedMTLDevice::Serialise_mtlDevice_newDefaultLibrary(SerialiserType &ser,
+                                                             WrappedMTLDevice *device,
+                                                             WrappedMTLLibrary *library,
+                                                             const void *pData, uint32_t bytesCount)
+{
+  SERIALISE_ELEMENT_LOCAL(Library, GetResID(library)).TypedAs("MTLLibrary"_lit);
```

Comment:
Is this the correct approach or should I serialization support to WrappedMTLLibrary and its implementation would serialize its resource ID ?

Reply:
In general case when referring to existing objects, yes this should be serialised via its ResourceId as a native serialisable type. However for creation functions there's a chicken-and-egg problem - on replay we haven't created the object yet because this is the function where it's created, so the object can't be serialised that way. Instead for create-time we do serialise the ResourceId in the way you have here so that's fine.